File: 0328-odd-even-linked-list/0328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:
            return head
        
        even = head
        odd_start = head.next
        odd = odd_start
        
        index = 2
        cur = head.next.next
        while even != None and odd != None:
            if even.next != None:
                even.next = even.next.next
                even = even.next
            if odd.next != None:    
                odd.next = odd.next.next
                odd = odd.next
        
        cur = head
        while cur.next != None:
            cur = cur.next
        
        cur.next = odd_start
        return head

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* oddEvenList(struct ListNode* head){
    struct ListNode *odd, *even, *evenh, *n;
    
    if (head == NULL || head->next == NULL)
        return head;
    
    odd = head;
    evenh = even = head->next;
    n = head->next->next;
    
    for (int i = 1; n != NULL; i++) {
        if (i % 2 == 0) {
            even->next = n;
            even = even->next;
        } else {
            odd->next = n;
            odd = odd->next;
        }
        n = n->next;
    }
    odd->next = evenh;
    even->next = NULL;
    return head;
}